test/gtest/codegen_group_conv.hpp

```diff
@@ -0,0 +1,567 @@
+                {1, 1  , 1 , 1 ,{8 , 8 , 8 }, {2, 2, 2}, {0, 0, 0}, {1, 1, 1}, {1, 1, 1}},
+                {1, 1  , 1 , 1 ,{8 , 8 , 8 }, {2, 2, 2}, {0, 0, 0}, {2, 2, 2}, {1, 1, 1}},
+                {1, 64 , 32, 16,{28, 28, 28}, {3, 3, 3}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+                {8, 128, 16, 32,{28, 28, 28}, {3, 3, 3}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+                {8, 128, 16, 16,{28, 28, 28}, {3, 3, 3}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+                {4, 128, 8 , 4 ,{28, 28, 28}, {2, 2, 2}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+                {4, 128, 4 , 8 ,{28, 28, 28}, {3, 3, 3}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+                {2, 128, 2 , 2 ,{28, 28, 28}, {3, 3, 3}, {1, 1, 1}, {1, 1, 1}, {1, 1, 1}},
+            };
+            // clang-format on
+        }
+        else
+        {
+            std::abort();
+        }
+    }
+};
+
+template <unsigned NDIM, typename T, Direction CONV_DIR, bool cg>
+struct GroupConvTestFix
+    : public ::testing::TestWithParam<
+          std::tuple<GroupConvTestConfig<NDIM>, float, float, miopenTensorLayout_t>>
+{
+    static_assert(NDIM == 2u || NDIM == 3u, "NDIM must be 2 for 2D Conv and 3 for 3D Conv");
+
+private:
+    using Base = ::testing::TestWithParam<
+        std::tuple<GroupConvTestConfig<NDIM>, float, float, miopenTensorLayout_t>>;
+
+    template <typename F>
+    void SetupFwd(F&& gen_value)
+    {
+        input.generate(gen_value);
+        weights.generate(gen_value);
+        std::fill(output.begin(), output.end(), T(0));
+    }
+
+    template <typename F>
+    void SetupBwd(F&& gen_value)
+    {
+        output.generate(gen_value);
+        weights.generate(gen_value);
+        std::fill(input.begin(), input.end(), T(0));
+    }
+
+    template <typename FI, typename FO>
+    void SetupWrw(FI&& gen_value_in, FO&& gen_value_out)
+    {
+        input.generate(gen_value_in);
+        output.generate(gen_value_out);
+        std::fill(weights.begin(), weights.end(), T{0});
+    }
+
+    void verify(const tensor<T>& computed)
+    {
+        EXPECT_FALSE(miopen::range_zero(ref)) << "Cpu data is all zeros";
+        EXPECT_FALSE(miopen::range_zero(computed)) << "Gpu data is all zeros";
+        EXPECT_TRUE(miopen::range_distance(ref) == miopen::range_distance(computed));
+
+        /// \todo figure out a better threshold for error checking, esp. for bwd
+        /// data and weight passes. --amberhassaan
+        double threshold = 80;
+        if(CONV_DIR == Direction::Forward)
+        {
+            threshold *= std::numeric_limits<T>::epsilon();
+        }
+        else
+        {
+            threshold = 3.0e-3;
+        }
+        auto error = miopen::rms_range(ref, computed);
+
+        EXPECT_FALSE(miopen::find_idx(ref, miopen::not_finite) >= 0)
+            << "Non finite number found in the reference output";
+
+        EXPECT_TRUE(error <= threshold)
+            << "Error beyond tolerance Error:" << error << ",  Threshold: " << threshold;
```

Comment:
```suggestion
        EXPECT_LE(error, threshold);
```
It will print something like:
```
Mismatch, error must be less than threshold :
error = 100500
threshold = 10
```